plugins/resource_monitor_plugin/resource_monitor_plugin.cpp

```diff
@@ -0,0 +1,157 @@
+/**
+   It was reported from a customer that when file system which 
+   "data/blocks" belongs to is running out of space, the producer  
+   continued to produce blocks and update state but the blocks log was
+   "corrupted" in that it no longer contained all the irreversible blocks.
+   It was also observed that when file system which "data/state"
+   belons to is running out of space, nodeos will crash with SIGBUS as
+   the state file is unable to acquire new pages.
+
+   The solution is to have a dedicated plugin to monitor resource 
+   usages (file system space now, CPU, memory, and networking
+   bandwidth in the future).
+   The plugin uses a thread to periodically check space usage of file
+   systems of directories being monitored. If space used
+   is over a predefined threshold, a graceful shutdown is initiated.
+**/
+
+#include <eosio/resource_monitor_plugin/resource_monitor_plugin.hpp>
+#include <eosio/resource_monitor_plugin/file_space_handler.hpp>
+#include <eosio/resource_monitor_plugin/system_file_space_provider.hpp>
+
+#include <eosio/chain/exceptions.hpp>
+
+#include <fc/exception/exception.hpp>
+#include <fc/log/logger_config.hpp> // set_os_thread_name
+
+#include <boost/filesystem.hpp>
+
+#include <thread>
+
+#include <sys/stat.h>
+
+using namespace eosio::resource_monitor;
+
+namespace bfs = boost::filesystem;
+
+namespace eosio {
+   static appbase::abstract_plugin& _resource_monitor_plugin = app().register_plugin<resource_monitor_plugin>();
+
+   static vector<bfs::path> directories_monitored;
+
+class resource_monitor_plugin_impl {
+public:
+   resource_monitor_plugin_impl()
+      :space_handler(system_file_space_provider(), ctx)
+      {
+      }
+
+   void set_program_options(options_description&, options_description& cfg) {
+      cfg.add_options()
+         ( "resource-monitor-interval-seconds", bpo::value<uint32_t>()->default_value(def_sleep_time_in_secs),
+           "Time in seconds between two consecutive checks of resource usage. Should be greater than 1 and less than 60 (one minute)" )
+         ( "resource-monitor-space-threshold", bpo::value<uint32_t>()->default_value(def_used_space_threshold_in_percentage),
+             "Threshold in terms of percentage of used space vs total space. If used space is over the threshold, a graceful shutdown is initiated. The value should be greater than 1 to less than 99" )
+         ;
+   }
+
+   void plugin_initialize(const appbase::variables_map& options) {
+      ilog("plugin_initialize");
```

Comment:
I think this should be debug level logging.

Reply:
Agreed.